src/libopenrave/libopenrave.cpp

```diff
@@ -2292,9 +2292,6 @@ void TriMesh::Append(const TriMesh& mesh)
 {
     int offset = (int)vertices.size();
     vertices.insert(vertices.end(), mesh.vertices.begin(), mesh.vertices.end());
-    if( indices.capacity() < indices.size()+mesh.indices.size() ) {
```

Comment:
@rschlaikjer why the pre-allocation is removed?

Reply:
This kind of reserve is an antipattern, especially the way that it plays into the larger call scope here - IIRC this function is called repeatedly, and with this explicit reserve
- You force a realloc, but more importantly
- You force that realloc _to only be as large as the new items added this cycle_.

If a vector internally reallocs during an insert, it will resize exponentially, so you'll end up with order log2(n) reallocs. If you force it to reserve on each call, you will end up doing n reallocs instead.

Reply:
How about this, then `indices` should know how much it should allocate before hand:

```cpp
#include <boost/range/adaptor/transformed.hpp>

void TriMesh::Append(const TriMesh& mesh)
{
    int offset = (int)vertices.size();
    vertices.insert(vertices.end(), mesh.vertices.begin(), mesh.vertices.end());
    auto indicesWithOffset = mesh.indices | boost::adaptors::transformed([&](const int &indice){return indice+offset;});
    indices.insert(indices.end(), indicesWithOffset.begin(), indicesWithOffset.end());
}
```

(uh-oh, there exists auto...)

Reply:
oh

why don't you use indices.resize() like L2303 vertices, then reserve happens automatically

Reply:
Both of these attempts to outsmart std::vector's internal logic seem to be slower than just letting it grow based on its internal heuristics. Simplified test case:
```c++
int main()
{
    std::vector<long> v;
    for (long i = 0; i < ITERCOUNT i++) {
#ifdef RESERVE
        v.reserve(v.size() + 1);
#elif defined RESIZE
        v.resize(v.size() + 1);
#endif
        v.emplace_back(i);
    }
}
```
With itercount = 100'000, it's clear just how pessimal reserve is:
```c++
mujin@G6DC229006MN:/h/m/cpp_units$ g++ -O3 improper_reserve.cpp && time ./a.out && g++ -O3 -DRESERVE improper_reserve.cpp && time ./a.out && g++ -O3 -DRESIZE improper_reserve.cpp && time ./a.out
// simple
real	0m0.002s
user	0m0.000s
sys	0m0.002s
// reserve
real	0m5.141s
user	0m0.964s
sys	0m4.177s
// resize
real	0m0.002s
user	0m0.002s
sys	0m0.000s
```

If we change itercount to 1'000'000'000 and ignore reserve because it would be too slow to complete, it seems like even resize isn't as fast as just calling emplace back:
```c++
mujin@G6DC229006MN:/h/m/cpp_units$ g++ -O3 improper_reserve.cpp && time ./a.out  && g++ -O3 -DRESIZE improper_reserve.cpp && time ./a.out
// simple
real	0m3.242s
user	0m0.869s
sys	0m2.374s
// resize
real	0m7.839s
user	0m2.399s
sys	0m5.438s

```

Reply:
Oops, sorry, flaw in my test case for resize there. Leaving it for posterity, but with correct semantics
```c++
int main()
{
    std::vector<long> v;
    for (long i = 0; i < 1'000'000'000; i++) {
#ifdef RESERVE
        v.reserve(v.size() + 1);
        v.emplace_back(i);
#elif defined RESIZE
        v.resize(v.size() + 1);
        v.back() = i;
#else
        v.emplace_back(i);
#endif
    }
}
```
resize still seems to be slower than the naive implementation
```c++
mujin@G6DC229006MN:/h/m/cpp_units$ g++ -O3 improper_reserve.cpp && time ./a.out  && g++ -O3 -DRESIZE improper_reserve.cpp && time ./a.out
// simple
real	0m3.331s
user	0m0.876s
sys	0m2.455s
// resize
real	0m3.516s
user	0m1.196s
sys	0m2.320s
```

Reply:
Let's try more practical case...

```
#include <vector>
#include <stdlib.h>

int main(int argc, char **argv)
{
    int step = strtol(argv[1], NULL, 10);
    std::vector<long> v;
    for (long i = 0; i < 1'000'000'000; i++) {
#ifdef RESERVE
        if(i%step==0)v.reserve(v.size() + step);
        v.emplace_back(i);
#elif defined RESIZE
        if(i%step==0)v.resize(v.size() + step);
        v[i] = i;
#else
        if(i%step>=0)v.emplace_back(i);
#endif
    }
}
```

```
$ g++ -O3 improper_reserve.cpp && time ./a.out 1000 && g++ -O3 -DRESIZE improper_reserve.cpp && time ./a.out 1000
// simple
real	0m13.288s
user	0m9.170s
sys	0m4.111s
// resize
real	0m12.503s
user	0m8.279s
sys	0m4.223s
```

Reply:
There seems to be a bug in your test case as written where it won't emplace the same amount of elements for the final case (should just be no if check). Interestingly, I get results that prefer the simple case even when fixed:
```c++
#include <stdlib.h>
#include <stdio.h>

#include <vector>

int main(int argc, char** argv)
{
    int step = strtol(argv[1], NULL, 10);
    std::vector<long> v;
    for (long i = 0; i < 1'000'000'000; i++) {
#ifdef RESERVE
        if (i % step == 0)
            v.reserve(v.size() + step);
        v.emplace_back(i);
#elif defined RESIZE
        if (i % step == 0)
            v.resize(v.size() + step);
        v[i] = i;
#else
        v.emplace_back(i);
#endif
    }
    fprintf(stderr, "Total vector size: %u\n", v.size());
}
```

```
g++ -O3 foo.cpp && time ./a.out 1000 && g++ -O3 -DRESIZE foo.cpp && time ./a.out 1000
Total vector size: 1000000000
// Naive
real	0m2.288s
user	0m0.780s
sys	0m1.508s
Total vector size: 1000000000
// Resize
real	0m4.161s
user	0m2.810s
sys	0m1.351s
````

Reply:
Confused by your result actually - thought it might be a cpu arch difference but ran the test on a controller and saw the same skew I get locally
```
root@controller1007:~# g++ -O3 foo.cpp && time ./a.out 1000 && g++ -O3 -DRESIZE foo.cpp && time ./a.out 1000
Total vector size: 1000000000

real	0m5.908s
user	0m1.992s
sys	0m3.915s
Total vector size: 1000000000

real	0m13.029s
user	0m8.849s
sys	0m4.180s
```

Reply:
@rschlaikjer because you deleted the division...

Reply:
Ah, I see what you were getting at. Still seems to be about margin of error, and of course without the resize call you don't _need_ the additional check...
```c++
#elif defined RESIZE
        if (i % step == 0)
            v.resize(v.size() + step);
        v[i] = i;
#else
        if (i % step == 0)
            asm volatile("nop");
        v.emplace_back(i);
```
```

mujin@G6DC229006MN:/h/mujin$ g++ -O3 foo.cpp && time ./a.out 1000 && echo $? && g++ -O3 -DRESIZE foo.cpp && time ./a.out 1000
Total vector size: 1000000000

real	0m4.135s
user	0m2.694s
sys	0m1.439s
0
Total vector size: 1000000000

real	0m4.149s
user	0m2.748s
sys	0m1.400s
```

Reply:
```cpp
#include <vector>
#include <stdlib.h>

int main(int argc, char **argv)
{
    int step = strtol(argv[1], NULL, 10);
    std::vector<long> v;
    for (long i = 0; i < 1'000'000'000; ) {
#if defined(RESIZE)
        v.resize(v.size() + step);
        for (long j = 0; j < step; ++j, ++i) {
            v[i] = i;
        }
#else
        v.emplace_back(i); ++i;
#endif
    }
}
```

```
$ g++ -O3 improper_reserve.cpp && time ./a.out 1000 && g++ -O3 -DRESIZE improper_reserve.cpp && time ./a.out 1000

real	0m8.288s
user	0m2.634s
sys	0m5.619s

real	0m7.148s
user	0m2.389s
sys	0m4.758s
```

Reply:
Running that same test case on c1007
```
root@controller1007:~# g++ -O3 foo.cpp && time ./a.out 1000 && g++ -O3 -DRESIZE foo.cpp && time ./a.out 1000

real	0m5.705s
user	0m1.688s
sys	0m4.016s

real	0m5.803s
user	0m1.554s
sys	0m4.241s
```

The difference here is again below the noise floor. At the end of the day, I think that the use case for explicit `resize` (or reserve) is that you know the _final_ size of the container. Resizing speculatively is more of a gamble - if your resize is large relative to the current size of the vector, it will be advantageous (doubling interval is small, you will save some reallocs). If your resize is small relative to the current size of the vector, it's just overhead (doubling interval already large, resize already within cap).

Reply:
@ziyan please decide which, simple emplace_back or resize.

Reply:
I think this is effectively resolved by the change to resize + assign as per Rosen's suggestion.